services/datamanager/builtin/capture_test.go

```diff
@@ -235,13 +199,121 @@ func TestDataCaptureEnabled(t *testing.T) {
+
+	waitForCaptureFilesToExceedN(captureDir, 0)
+	testFilesContainSensorData(t, captureDir, uniqueDataValues)
+
+	cancelPassTime1()
+	<-donePassingTime1
+
+	// Change resource on robot.
+	for resource := range resources {
+		if resource.Name == "arm1" {
+			newResource := inject.NewArm(resource.Name)
+			newResource.EndPositionFunc = func(ctx context.Context, extra map[string]interface{}) (spatialmath.Pose, error) {
+				return spatialmath.NewPoseFromPoint(r3.Vector{X: 888, Y: 888, Z: 888}), nil
+			}
+			resources[resource] = newResource
+		}
+	}
+
+	err = dmsvc.Reconfigure(context.Background(), resources, resource.Config{
+		ConvertedAttributes: config,
+	})
+	test.That(t, err, test.ShouldBeNil)
+
+	passTimeCtx2, cancelPassTime2 := context.WithCancel(context.Background())
+	donePassingTime2 := passTime(passTimeCtx2, mockClock, captureInterval)
+
+	// Test that sensor data is captured from the new collector.
+	waitForCaptureFilesToExceedN(captureDir, len(getAllFileInfos(captureDir)))
+	testFilesContainSensorData(t, captureDir, uniqueDataValues+1)
+
+	cancelPassTime2()
+	<-donePassingTime2
+}
+
+// passTime repeatedly increments mc by interval until the context is canceled.
+func passTime(ctx context.Context, mc *clk.Mock, interval time.Duration) chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				close(done)
+				return
+			default:
+				mc.Add(interval)
+			}
+		}
+	}()
+	return done
+}
+
+// testFilesContainSensorData verifies that the files in `dir` contain sensor data, and that there are `expectedUniqueDataValues` num of unique datapoints.
+func testFilesContainSensorData(t *testing.T, dir string, expectedUniqueDataValues int) {
+	t.Helper()
+	var sd []*v1.SensorData
+	valueSet := map[float64]struct{}{}
+	filePaths := getAllFilePaths(dir)
+	for _, path := range filePaths {
+		d, err := datacapture.SensorDataFromFilePath(path)
+		// It's possible a file was closed (and so its extension changed) in between the points where we gathered
+		// file names and here. So if the file does not exist, check if the extension has just been changed.
+		if errors.Is(err, os.ErrNotExist) {
+			path = strings.TrimSuffix(path, filepath.Ext(path)) + datacapture.FileExt
+			d, err = datacapture.SensorDataFromFilePath(path)
+		}
+		for _, datapoint := range d {
+			valueSet[datapoint.GetStruct().GetFields()["Number"].GetStructValue().GetFields()["Dual"].GetStructValue().GetFields()["Jmag"].GetNumberValue()] = struct{}{}
+		}
+		test.That(t, err, test.ShouldBeNil)
+		sd = append(sd, d...)
+	}
+
+	test.That(t, len(sd), test.ShouldBeGreaterThan, 0)
+	for _, d := range sd {
+		test.That(t, d.GetStruct(), test.ShouldNotBeNil)
+		test.That(t, d.GetMetadata(), test.ShouldNotBeNil)
+	}
+
+	// Each collector's mocked capture method should output a different value. Assert that we see the expected number of unique data values.
+	test.That(t, len(valueSet), test.ShouldEqual, expectedUniqueDataValues)
```

Comment:
A suggestion (that you are free to reject): this validation feels pretty specific to the test you just added. The existing tests just cared about whether or not sensor data was written and that it was non-empty. I think the test validation would be clearer if `testFilesContainSensorData` accepted a `validateSensorData` function as one of its arguments so that each test can perform the specific validation that it needs. It would make it more clear what `TestSwitchResource` is actually checking. I'm thinking something like this:

```
func testFilesContainSensorData(t *testing.T, dir string, validateSensorData func(*testing.T, []*v1.SensorData))
```

Reply:
That makes sense, done!